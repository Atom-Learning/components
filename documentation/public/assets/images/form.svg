<svg width="217" height="195" fill="none" xmlns="http://www.w3.org/2000/svg"><rect width="59" height="11" rx="5.5" fill="#545454"/><rect x=".37" y="20.26" width="216.26" height="34.81" rx="2.59" fill="#fff"/><rect x="8.89" y="32.11" width="65.17" height="11.11" rx="5.55" fill="#CECECE"/><rect x=".37" y="20.26" width="216.26" height="34.81" rx="2.59" stroke="#545454" stroke-width=".74"/><rect y="79.19" width="57.03" height="11.11" rx="5.55" fill="#545454"/><rect x=".37" y="99.56" width="216.26" height="34.81" rx="2.59" fill="#fff"/><rect x="8.89" y="111.41" width="65.17" height="11.11" rx="5.55" fill="#CECECE"/><rect x=".37" y="99.56" width="216.26" height="34.81" rx="2.59" stroke="#545454" stroke-width=".74"/><path d="M0 161.45a2.96 2.96 0 0 1 2.96-2.96h106.65a2.96 2.96 0 0 1 2.96 2.96v29.63a2.96 2.96 0 0 1-2.96 2.96H2.96A2.96 2.96 0 0 1 0 191.08v-29.63Z" fill="#545454"/><rect x="23.7" y="170.71" width="65.17" height="11.11" rx="5.55" fill="#fff"/></svg>