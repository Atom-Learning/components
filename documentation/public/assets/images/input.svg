<svg width="293" height="40" viewBox="0 0 293 40" fill="none" xmlns="http://www.w3.org/2000/svg">
<rect x="0.5" y="0.5" width="292" height="39" rx="3.5" fill="white"/>
<path d="M12.7727 15.6136V14.3636H21.5V15.6136H17.8409V26H16.4318V15.6136H12.7727ZM25.8864 26.1818C25.0455 26.1818 24.3201 25.9962 23.7102 25.625C23.1042 25.25 22.6364 24.7273 22.3068 24.0568C21.9811 23.3826 21.8182 22.5985 21.8182 21.7045C21.8182 20.8106 21.9811 20.0227 22.3068 19.3409C22.6364 18.6553 23.0947 18.1212 23.6818 17.7386C24.2727 17.3523 24.9621 17.1591 25.75 17.1591C26.2045 17.1591 26.6534 17.2348 27.0966 17.3864C27.5398 17.5379 27.9432 17.7841 28.3068 18.125C28.6705 18.4621 28.9602 18.9091 29.1761 19.4659C29.392 20.0227 29.5 20.7083 29.5 21.5227V22.0909H22.7727V20.9318H28.1364C28.1364 20.4394 28.0379 20 27.8409 19.6136C27.6477 19.2273 27.3712 18.9223 27.0114 18.6989C26.6553 18.4754 26.2348 18.3636 25.75 18.3636C25.2159 18.3636 24.7538 18.4962 24.3636 18.7614C23.9773 19.0227 23.6799 19.3636 23.4716 19.7841C23.2633 20.2045 23.1591 20.6553 23.1591 21.1364V21.9091C23.1591 22.5682 23.2727 23.1269 23.5 23.5852C23.7311 24.0398 24.0511 24.3864 24.4602 24.625C24.8693 24.8598 25.3447 24.9773 25.8864 24.9773C26.2386 24.9773 26.5568 24.928 26.8409 24.8295C27.1288 24.7273 27.3769 24.5758 27.5852 24.375C27.7936 24.1705 27.9545 23.9167 28.0682 23.6136L29.3636 23.9773C29.2273 24.4167 28.9981 24.803 28.6761 25.1364C28.3542 25.4659 27.9564 25.7235 27.483 25.9091C27.0095 26.0909 26.4773 26.1818 25.8864 26.1818ZM32.3054 17.2727L34.3963 20.8409L36.4872 17.2727H38.0327L35.2145 21.6364L38.0327 26H36.4872L34.3963 22.6136L32.3054 26H30.7599L33.5327 21.6364L30.7599 17.2727H32.3054ZM43.7415 17.2727V18.4091H39.2188V17.2727H43.7415ZM40.5369 15.1818H41.8778V23.5C41.8778 23.8788 41.9328 24.1629 42.0426 24.3523C42.1563 24.5379 42.3002 24.6629 42.4744 24.7273C42.6525 24.7879 42.84 24.8182 43.0369 24.8182C43.1847 24.8182 43.3059 24.8106 43.4006 24.7955C43.4953 24.7765 43.571 24.7614 43.6278 24.75L43.9006 25.9545C43.8097 25.9886 43.6828 26.0227 43.5199 26.0568C43.357 26.0947 43.1506 26.1136 42.9006 26.1136C42.5218 26.1136 42.1506 26.0322 41.7869 25.8693C41.4271 25.7064 41.1278 25.4583 40.8892 25.125C40.6544 24.7917 40.5369 24.3712 40.5369 23.8636V15.1818Z" fill="#1F1F1F"/>
<rect x="0.5" y="0.5" width="292" height="39" rx="3.5" stroke="#545454"/>
</svg>
