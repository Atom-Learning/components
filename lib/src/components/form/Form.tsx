import React from 'react'
import {
  DefaultValues,
  FieldErrors,
  FieldValues,
  FormProvider,
  Mode,
  useForm,
  UseFormMethods
} from 'react-hook-form'

import { styled } from '~/stitches'

const StyledForm = styled('form', {})

type StyledFormProps = Omit<
  React.ComponentPropsWithoutRef<typeof StyledForm>,
  'onSubmit' | 'onError'
>

interface FormProps<TFormData extends FieldValues> extends StyledFormProps {
  defaultValues?: DefaultValues<TFormData>
  validationMode?: Mode
  onSubmit: (data: TFormData) => void | any
  onError?: (errors: FieldErrors<TFormData>) => void
  children:
    | React.ReactNode
    | ((methods: UseFormMethods<TFormData>) => React.ReactNode)
}

export const Form = <TFormData extends FieldValues>(
  props: FormProps<TFormData>
) => {
  const {
    children,
    defaultValues,
    validationMode = 'onBlur',
    onSubmit,
    onError,
    ...rest
  } = props

  const methods = useForm<TFormData>({
    defaultValues,
    mode: validationMode
  })

  return (
    <FormProvider {...methods}>
      <StyledForm
        aria-label="form"
        onSubmit={methods.handleSubmit(onSubmit, onError)}
        {...rest}
      >
        {typeof children === 'function' ? children(methods) : children}
      </StyledForm>
    </FormProvider>
  )
}

Form.displayName = 'Form'
