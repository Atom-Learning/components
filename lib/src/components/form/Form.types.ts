import type { Mode, UseFormMethods } from 'react-hook-form'

export enum StorageEnum {
  LOCAL = 'local',
  SESSION = 'session'
}

type ExcludeIncludeConfig = {
  exclude?: string[]
  include?: string[]
}

export type PersistOptions = {
  id: string
  storage?: StorageEnum
} & ExcludeIncludeConfig

export type FormPersistParams = {
  storage: Storage
} & ExcludeIncludeConfig

export type PersistFormWrapperValues = {
  persist: PersistOptions
  watch
  setValue
  children: React.ReactNode | any
}

export type FormContentValues = {
  formMethods
  handleSubmit: (
    submitHandler: (data: any) => void | any,
    submitErrorHandler?: (errors: any) => void
  ) => (e: any) => Promise<void>
  onSubmit: (data: any) => void | any
  onError?: (errors: any) => void
  children: React.ReactNode | any
}

export type FormValues = {
  defaultValues?: { [key: string]: string | number }
  onSubmit: (data: any) => void | any
  onError?: (errors: any) => void
  criteriaMode?: 'firstError' | 'all'
  validationMode?: Mode
  persist?: PersistOptions
} & (
  | { children: React.ReactNode; render?: never }
  | { children?: never; render: (methods: UseFormMethods) => React.ReactNode }
)
