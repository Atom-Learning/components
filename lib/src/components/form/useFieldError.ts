import * as React from 'react'
import delve from 'dlv'
import { useFormContext } from 'react-hook-form'

export const useFieldError = (
  fieldName: string,
  validationMode: 'firstError' | 'all' = 'firstError'
) => {
  const { errors } = useFormContext()

  const fieldErrors = getFieldErrors(errors, fieldName, validationMode)

  return fieldErrors
}

const getFieldErrors = (
  errors,
  fieldName,
  validationMode
): Array<string> | null => {
  // in case of migration to a newer version of react-hook-form the path needs to updated to account for dynamic fields naming convention
  const fieldPath = fieldName.split(/[,[\].]+?/).filter(Boolean)

  // const getErrors = () => delve(errors, fieldPath)
  const error = delve(errors, fieldPath)

  // `message` is the first validation error, as prioritised by react-hook-form.
  // `types` is an object containing all current validation errors, keyed by the validation type
  const { message, types } = error || {}

  if (!types && !message) return null

  if (validationMode === 'firstError' && message) return [message]

  return Object.values(types)
}
