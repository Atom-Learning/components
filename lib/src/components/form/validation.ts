export type ValidationOptions = {
  required?: Option<boolean> | string
  min?: Option<number>
  max?: Option<number>
  criteriaMode?: 'all' | 'firstError'
  minLength?: Option<number>
  maxLength?: Option<number>
  pattern?: Option<RegExp>
  validate?: Validator | { [key: string]: Validator }
  valueAsNumber?: boolean
  valueAsDate?: boolean
  process?: Processor
}

export type ValidationError = {
  message: string
  type:
    | 'required'
    | 'min'
    | 'max'
    | 'minLength'
    | 'maxLength'
    | 'pattern'
    | 'validate'
}

// Returns `true` if a value passes validation, or a string with the error if not
type Validator = (value: any) => boolean | string

// Processes a value after validation
type Processor = (value: any) => any

type Option<T> = {
  value: T
  message: string
}
