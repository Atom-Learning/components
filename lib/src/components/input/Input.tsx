import * as React from 'react'

import { styled } from '~/stitches'
import { disabledStyle } from '~/utilities'
import { Override } from '~/utilities/types'

const StyledInput = styled('input', {
  appearance: 'none',
  border: '1px solid $tonal400',
  borderRadius: '$0',
  boxShadow: 'none', // prevent default iOS default styling
  boxSizing: 'border-box',
  color: '$tonal600',
  cursor: 'text',
  display: 'block',
  fontFamily: '$body',
  px: '$3',
  transition: 'all 100ms ease-out',
  width: '100%',
  '&:focus': {
    borderColor: '$primary800',
    outline: 'none'
  },
  '&[disabled]': disabledStyle,
  '&::placeholder': {
    color: '$tonal300',
    opacity: 1
  },
  variants: {
    size: {
      sm: {
        height: '$3',
        fontSize: '$sm',
        lineHeight: 1.7
      },
      md: {
        height: '$4',
        fontSize: '$md',
        lineHeight: 2
      },
      lg: {
        height: '$5',
        fontSize: '$md',
        lineHeight: 2
      },
      xl: {
        height: '$6',
        fontSize: '$lg',
        lineHeight: 2
      }
    },
    state: {
      error: {
        border: '1px solid $danger'
      }
    }
  }
})

// override default 'type' property to prevent Input from being used to render
// checkboxes, radios etc â€” we have dedicated components for them
export type InputProps = Override<
  React.ComponentProps<typeof StyledInput>,
  {
    name: string
    as?: never
    type?: 'text' | 'number' | 'email' | 'password' | 'tel' | 'url' | 'search'
  }
>

export const Input: React.ForwardRefExoticComponent<InputProps> =
  React.forwardRef(({ type = 'text', size = 'md', ...rest }, ref) => {
    if (type === 'number') {
      return (
        <StyledInput
          type="text"
          inputMode="numeric"
          pattern="[0-9]*"
          size={size}
          {...rest}
          ref={ref}
        />
      )
    }

    return <StyledInput type={type} size={size} {...rest} ref={ref} />
  })

Input.displayName = 'Input'
