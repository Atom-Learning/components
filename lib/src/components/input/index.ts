export type { InputProps } from './Input'
export { Input, InputBackground, InputText } from './Input'
