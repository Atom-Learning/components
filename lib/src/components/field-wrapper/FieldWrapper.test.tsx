import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { axe } from 'jest-axe'
import * as React from 'react'

import { Input } from '../input'
import { FieldWrapper } from '.'
import { Form } from '../form'

const ExampleField = () => (
  // Form is required because the FieldWrapper uses the useFormContext hook
  <Form onSubmit={jest.fn()}>
    <FieldWrapper
      label="Example Field"
      fieldId="example"
      prompt={{ label: 'Example prompt', link: 'https://example.com' }}
    >
      <Input name="example" id="example" />
    </FieldWrapper>
  </Form>
)

describe('FieldWrapper component', () => {
  it('renders', async () => {
    const { container } = render(<ExampleField />)

    expect(container).toMatchSnapshot()
  })

  it('has no programmatically detectable a11y issues', async () => {
    const { container } = render(<ExampleField />)

    expect(await axe(container)).toHaveNoViolations()
  })

  it('links the label to an input', async () => {
    render(<ExampleField />)

    userEvent.click(await screen.findByText('Example Field'))

    expect(await screen.findByRole('textbox')).toHaveFocus()
  })

  // TODO: unskip when we can pass custom messages into the FieldWrapper
  it.skip('renders provided error', async () => {
    render(<ExampleField />)

    expect(await screen.findByText('Example error')).toBeVisible()
  })

  it('renders provided prompt', async () => {
    render(<ExampleField />)
    const link = await screen.findByRole('link')

    expect(link).toHaveTextContent('Example prompt')
    expect(link).toHaveAttribute('href', 'https://example.com')
  })
})
