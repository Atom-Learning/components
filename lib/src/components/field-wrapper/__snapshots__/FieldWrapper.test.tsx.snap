// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldWrapper component renders 1`] = `
@media  {
  .c-dhzjXW {
    display: flex;
  }

  .c-cVSpVL {
    color: var(--colors-tonal500);
    font-family: var(--fonts-body);
    margin: 0;
  }

  .c-kgcOue {
    background: unset;
    border: unset;
    padding: unset;
    color: var(--colors-primary);
    cursor: pointer;
    font-family: var(--fonts-body);
    text-decoration: none;
  }

  .c-kgcOue:focus,
  .c-kgcOue:hover {
    color: var(--colors-primaryMid);
    text-decoration: underline;
  }

  .c-kgcOue:active {
    color: var(--colors-primaryDark);
  }

  .c-fIXJvv > .c-kgcOue,
  .c-iSnHBU > .c-kgcOue,
  .c-PJLV > .c-kgcOue {
    font-size: 100%;
    line-height: 1;
  }

  .c-fIXJvv > .c-kgcOue::before,
  .c-fIXJvv > .c-kgcOue::after,
  .c-iSnHBU > .c-kgcOue::before,
  .c-iSnHBU > .c-kgcOue::after,
  .c-PJLV > .c-kgcOue::before,
  .c-PJLV > .c-kgcOue::after {
    content: none;
  }

  .c-fZTBsJ {
    -webkit-appearance: none;
    appearance: none;
    border: 1px solid var(--colors-tonal400);
    border-radius: var(--radii-0);
    box-shadow: none;
    box-sizing: border-box;
    color: var(--colors-tonal600);
    cursor: text;
    display: block;
    font-family: var(--fonts-body);
    padding-left: var(--space-3);
    padding-right: var(--space-3);
    transition: all 100ms ease-out;
    width: 100%;
  }

  .c-fZTBsJ:focus {
    border-color: var(--colors-primary);
    outline: none;
  }

  .c-fZTBsJ[disabled] {
    background-color: var(--colors-tonal100);
    color: var(--colors-tonal400);
    cursor: not-allowed;
  }

  .c-fZTBsJ::placeholder {
    color: var(--colors-tonal300);
    opacity: 1;
  }

  .c-dbrbZt {
    display: inline-block;
    fill: none;
    stroke: currentcolor;
    stroke-linecap: round;
    stroke-linejoin: round;
    vertical-align: middle;
  }

  .c-fzYfTW {
    flex-shrink: 0;
    margin-right: var(--space-2);
  }

  .c-fIXJvv {
    color: var(--colors-tonal600);
    font-family: var(--fonts-body);
    font-weight: 400;
    margin: 0;
  }

  .c-fIXJvv > .c-fIXJvv:before,
  .c-fIXJvv > .c-fIXJvv:after {
    display: none;
  }
}

@media  {
  .c-cVSpVL-gvmVBy-size-md {
    font-size: var(--fontSizes-md);
    line-height: 1.5;
  }

  .c-cVSpVL-gvmVBy-size-md::before {
    content: '';
    margin-bottom: -0.3864em;
    display: table;
  }

  .c-cVSpVL-gvmVBy-size-md::after {
    content: '';
    margin-top: -0.3864em;
    display: table;
  }

  .c-cVSpVL-grKRUr-type-block {
    display: block;
    font-weight: 600;
  }

  .c-kgcOue-bndJoy-size-sm {
    font-size: var(--fontSizes-sm);
    line-height: 1.53;
  }

  .c-kgcOue-bndJoy-size-sm::before {
    content: '';
    margin-bottom: -0.4056em;
    display: table;
  }

  .c-kgcOue-bndJoy-size-sm::after {
    content: '';
    margin-top: -0.4056em;
    display: table;
  }

  .c-fZTBsJ-pHZal-size-md {
    height: var(--sizes-4);
    font-size: var(--fontSizes-md);
  }

  .c-PJLV-iymbh-theme-error {
    color: var(--colors-danger);
  }

  .c-dbrbZt-cyeZeh-size-sm {
    height: var(--sizes-1);
    width: var(--sizes-1);
    stroke-width: 1.5;
  }

  .c-fIXJvv-bndJoy-size-sm {
    font-size: var(--fontSizes-sm);
    line-height: 1.53;
  }

  .c-fIXJvv-bndJoy-size-sm::before {
    content: '';
    margin-bottom: -0.4056em;
    display: table;
  }

  .c-fIXJvv-bndJoy-size-sm::after {
    content: '';
    margin-top: -0.4056em;
    display: table;
  }
}

@media  {
  .c-dhzjXW-iwHglP-css {
    justify-content: space-between;
    align-items: center;
    margin-bottom: var(--space-3);
  }

  .c-dhzjXW-ijkHUQU-css {
    margin-top: var(--space-2);
  }

  .c-fIXJvv-igszWir-css {
    color: inherit;
    transform: translateY(var(--space-0));
  }
}

<div>
  <div
    class="c-PJLV"
  >
    <div
      class="c-dhzjXW c-dhzjXW-iwHglP-css"
    >
      <label
        class="c-cVSpVL c-cVSpVL-gvmVBy-size-md c-cVSpVL-grKRUr-type-block"
        for="example"
      >
        Example Field
      </label>
      <a
        class="c-kgcOue c-kgcOue-bndJoy-size-sm"
        href="https://example.com"
      >
        Example prompt
      </a>
    </div>
    <input
      class="c-fZTBsJ c-fZTBsJ-pHZal-size-md"
      id="example"
      name="example"
      type="text"
    />
    <div
      class="c-dhzjXW c-PJLV c-PJLV-iymbh-theme-error c-dhzjXW-ijkHUQU-css"
    >
      <svg
        aria-hidden="true"
        class="c-dbrbZt c-dbrbZt-cyeZeh-size-sm c-fzYfTW"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 8v5"
        />
        <line
          x1="12"
          x2="12"
          y1="16"
          y2="16"
        />
        <circle
          cx="12"
          cy="12"
          r="10"
        />
      </svg>
      <span
        class="c-fIXJvv c-fIXJvv-bndJoy-size-sm c-fIXJvv-igszWir-css"
      >
        Example error
      </span>
    </div>
  </div>
</div>
`;
